SIL linker cleanup

slavapestov opened the pull request.

slavapestov committed:
SIL: Remove outdated remark about transparent functions

Transparent function bodies are in fact available externally.

slavapestov committed:
SIL: Simplify vtable deserialization

Both SILModule::lookUpVTable() and SILLinker::linkVTable() were
doing the same thing. Try simplifying the latter.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke benchmark

slavapestov commented:
@swift-ci Please smoke test compiler performance

gottesmm commented on lib/SILOptimizer/Mandatory/MandatoryInlining.cpp:
```diff
@@ -663,15 +663,6 @@ class MandatoryInlining : public SILModuleTransform {
                                SetFactory, SetFactory.getEmptySet(), CHA);
     }
 
-    // Make sure that we de-serialize all transparent functions,
```
I just want to double check. What was the commit that fixed this?

slavapestov replied:
It wasn't a "fix" per se, just a removal of an unsound optimization. See f10f8d55dc43ad1addfc5e4854f691e0d0099049.

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -56,84 +56,10 @@ bool SILLinkerVisitor::processFunction(SILFunction *F) {
-///
-/// This method assumes that the caller made sure that no vtable existed in
-/// Mod.
-SILVTable *SILLinkerVisitor::processClassDecl(const ClassDecl *C) {
```
Why get rid of this?

slavapestov replied:
I don't believe it was doing anything, because SILModule::lookupVTable() was already loading the vtable if one was not found.

gottesmm commented on lib/SIL/Linker.cpp:
```diff
@@ -56,84 +56,10 @@ bool SILLinkerVisitor::processFunction(SILFunction *F) {
-
-  // If the SILModule does not have the VTable, attempt to deserialize the
-  // VTable. If we fail to do that as well, bail.
-  if (!Vtbl || !(Vtbl = Loader->lookupVTable(D->getName())))
```
And this?

gottesmm reviewed (commented):
Some questions. I just want to understand the overall flow of what you are changing here.

slavapestov committed:
SIL: Remove some uses of the -disable-sil-linking frontend flag

slavapestov committed:
wtable deser fix

slavapestov committed:
less eager vtable linking

slavapestov committed:
SIL: Centralize logic for linking function in SIL linker

In a few places we need to decide if we're going to add
a referenced function to the worklist. Factor this out,
and also add two new conditions in preparation for an
upcoming change:

- @_transparent functions should be linked in, regardless
  of visibility

- hidden_external functions should be linked in, because
  that is how declarations of PublicNonABI functions are
  deserialized, for example default argument generators

For now this is NFC because mandatory inlining eagerly
links in all referenced functions, but once that changes
the above logic is needed to generate valid code.

slavapestov committed:
SILOptimizer: Add 'mandatory SIL linker' pass

This pass only links in what is necessary for code generation, namely:

- @_transparent functions
- shared functions
- hidden_external functions

For now, this is NFC, because mandatory inlining eagerly links in
everything.

slavapestov committed:
SIL: Don't link vtables in LinkNormal mode

Mandatory inlining performs devirtualization, which links in
vtables and referenced functions.

slavapestov committed:
SILOptimizer: Check if a function is transparent before linking it

slavapestov committed:
IRGen: Tighten up assertions around foreign class metadata

ClassDecl::isForeign() is true for both CF types and runtime-only classes.
In some places, we do not expect to see the latter.

slavapestov committed:
IRGen: Fix protocol conformances for Objective-C runtime classes

We were emitting foreign type metadata for this case, because of an
invalid usage of ClassDecl::isForeign(), which also returns true
for runtime-only classes.

Fixes <rdar://problem/39117602>.

slavapestov committed:
IRGen: Fix another bad use of getAddrOfForeignTypeMetadataCandidate()

slavapestov commented:
This is obsolete
